url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/KlangUtil.java

```diff
@@ -0,0 +1,720 @@
+/**
+ * This file is distributed under a BSD-style license. See the included LICENSE.txt file
+ * for more information. 
+ * Copyright (c) 2009, James Halliday
+ * All rights reserved.
+ */
+package malictus.klang;
+
+import java.awt.*;
+import java.security.MessageDigest;
+import java.text.*;
+import java.util.zip.*;
+import java.io.*;
+import java.util.*;
+import malictus.klang.primitives.BadValueException;
+import malictus.klang.ui.KlangProgressWindow;
+
+/**
+ * General purpose utilities for Klang. Includes file utilites, string utilities, and 
+ * checksum utilities.
+ * @author Jim Halliday
+ */
+public class KlangUtil {
+
+	private KlangUtil() {}
+	
+	/**
+	 * Attempt to decompress some compressed data (using ZLIB)
+	 * @param compressed the compressed byte array
+	 * @return a new byte array with the data uncompressed
+	 */
+	public static byte[] inflate(byte[] compressed) throws BadValueException {
+		Inflater decompresser = new Inflater();
+		decompresser.setInput(compressed, 0, compressed.length);
+		//not the best way to do this!
+		//TODO fix
+		Vector<byte[]> results = new Vector<byte[]>();
+		boolean finished = false;
+		try {
+			while (finished == false) {
+				byte[] result = new byte[KlangConstants.BUFFER_SIZE];
+				int resultLength = decompresser.inflate(result);
+				if (resultLength == 0) {
+					finished = true;
+				} else if (resultLength < result.length) {
+					finished = true;
+					byte[] endone = new byte[resultLength];
+					System.arraycopy(result, 0, endone, 0, endone.length);
+					results.add(endone);
+				} else {
+					results.add(result);
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 43:52, 93:102.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.